### PR TITLE
[flutter_tool] Screenshot command must require device only for _kDeviceType

### PR DESCRIPTION
**Pull request by iskakaushik**
There are cases where we have access to the observatory without having a device connection accessible.

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/screenshot.dart [hunk 1]
```diff
@@ -66,8 +66,8 @@ class ScreenshotCommand extends FlutterCommand {
   @override
   Future<FlutterCommandResult> verifyThenRunCommand(String commandPath) async {
     device = await findTargetDevice();
-    if (device == null) {
-      throwToolExit('Must have a connected device');
+    if (argResults[_kType] == _kDeviceType && device == null) {
```
Comment (jason-simmons) [inline packages/flutter_tools/lib/src/commands/screenshot.dart]: Merge this and and the `device.supportsScreenshot` check into one `if (argResults[_kType] == _kDeviceType)` block

Comment (iskakaushik) [reply packages/flutter_tools/lib/src/commands/screenshot.dart]: The error message will be different in both the cases, no?

1. The first case is where we must have a device connected.
2. The device doesn't support screenshots even when connected.

Were you proposing nesting the `if` conditions inside the parent `if`?

Comment (jason-simmons) [reply packages/flutter_tools/lib/src/commands/screenshot.dart]: Yes

Comment (iskakaushik) [reply packages/flutter_tools/lib/src/commands/screenshot.dart]: Done.

**Comment by jonahwilliams** [comment]
Please add a test case for this new behavior

**Comment by iskakaushik** [comment]
> Please add a test case for this new behavior

Done

File: packages/flutter_tools/test/general.shard/screenshot_command_test.dart [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+import 'package:flutter_tools/src/base/common.dart';
+import 'package:flutter_tools/src/commands/screenshot.dart';
+import 'package:matcher/matcher.dart';
+
+import '../src/common.dart';
+import '../src/context.dart';
+
+void main() {
+  final Matcher expectToolError = throwsA(const TypeMatcher<ToolExit>());
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/screenshot_command_test.dart]: There should already be a matched called throwsToolExit

File: packages/flutter_tools/test/general.shard/screenshot_command_test.dart [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+  setUpAll(() {});
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/screenshot_command_test.dart]: remove empty setups?

Comment (iskakaushik) [reply packages/flutter_tools/test/general.shard/screenshot_command_test.dart]: i copied over the existing structure from `build_info_test.dart`, I should've realized the empty setups were a noop.

**Review by jonahwilliams — APPROVED** [review]
LGTM, thanks for the tests!

File: packages/flutter_tools/test/general.shard/screenshot_command_test.dart (hunk 2 above)
Comment (iskakaushik) [reply packages/flutter_tools/test/general.shard/screenshot_command_test.dart]: Switched to using this.

File: packages/flutter_tools/lib/src/commands/screenshot.dart [hunk 4]
```diff
@@ -63,18 +63,25 @@ class ScreenshotCommand extends FlutterCommand {
 
   Device device;
 
+  static void validateOptions(String screenshotType, Device device, String observatoryUri) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/screenshot.dart]: A switch statement might work better here:

```dart
switch (screenshotType) {
  case _kDeviceType:
    if (device == null) {
    }
    if (!device.supportsScreenshot) {
    }
  default:
    if (observatoryUri == null) {
    }
}
```

Comment (iskakaushik) [reply packages/flutter_tools/lib/src/commands/screenshot.dart]: switched :-)

**Review by zanderso — APPROVED** [review]

**Pull request merged**